File: skills/webhook_manager.py
import logging
from skills.base_skill import BaseSkill

logger = logging.getLogger(__name__)


class WebhookManagerSkill(BaseSkill):
    def __init__(self, manager):
        self.manager = manager

    @property
    def name(self) -> str:
        return "webhook_manager"

    @property
    def description(self) -> str:
        return "Verwaltet Webhooks: erstellen, auflisten, loeschen. Webhooks empfangen externe HTTP-Anfragen und loesen Ereignisse aus."

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "action": {
                    "type": "string",
                    "enum": ["create", "list", "delete"],
                    "description": "Aktion: create, list, oder delete",
                },
                "name": {
                    "type": "string",
                    "description": "Name des Webhooks (fuer create/delete)",
                },
                "description": {
                    "type": "string",
                    "description": "Beschreibung des Webhooks (nur bei create)",
                },
            },
            "required": ["action"],
        }

    async def execute(self, action: str, name: str = "", description: str = "", **kwargs) -> str:
        try:
            if action == "create":
                if not name:
                    return "Fehler: 'name' ist erforderlich."
                result = await self.manager.create(name, description)
                if "error" in result:
                    return result["error"]
                url = f"/api/webhooks/incoming/{result['name']}?token={result['token']}"
                return (
                    f"Webhook '{result['name']}' erstellt.\n"
                    f"URL: {url}\n"
                    f"Token: {result['token']}\n"
                    f"Nutze diese URL fuer externe HTTP POST-Anfragen."
                )

            elif action == "list":
                webhooks = await self.manager.list_all()
                if not webhooks:
                    return "Keine Webhooks vorhanden."
                lines = [
                    f"- **{w['name']}**: {w['description'] or '(keine Beschreibung)'} "
                    f"(Token: {w['token'][:8]}...)"
                    for w in webhooks
                ]
                return "Registrierte Webhooks:\n" + "\n".join(lines)

            elif action == "delete":
                if not name:
                    return "Fehler: 'name' ist erforderlich."
                return await self.manager.delete(name)

            else:
                return f"Unbekannte Aktion: {action}"

        except Exception as e:
            logger.exception("webhook_manager failed")
            return f"Fehler: {e}"

File: src/tools/builtin/webhook-manager.ts
import type { Tool } from '../registry.js'
import type { Database } from '../../infra/db.js'

export class WebhookManagerTool implements Tool {
  readonly name = 'webhook_manager'
  readonly description = 'Verwaltet eingehende Webhooks: Erstellen, Auflisten und Loeschen.'
  readonly parameters = {
    type: 'object',
    properties: {
      action: {
        type: 'string',
        enum: ['list', 'create', 'delete'],
        description: 'Die auszufuehrende Aktion',
      },
      name: { type: 'string', description: 'Name des Webhooks' },
      event_type: { type: 'string', description: 'Event-Typ der ausgeloest wird' },
      secret: { type: 'string', description: 'Geheimnis fuer Webhook-Validierung (optional)' },
    },
    required: ['action'],
  }

  constructor(private readonly db: Database) {}

  async execute(args: Record<string, unknown>): Promise<string> {
    const action = String(args.action ?? '')

    switch (action) {
      case 'list': {
        const hooks = this.db.listWebhooks()
        if (!hooks.length) return 'Keine Webhooks konfiguriert.'
        return hooks.map(h => `[${h.name}] Event: ${h.event_type} | URL: /api/webhooks/${h.name}`).join('\n')
      }

      case 'create': {
        const name = String(args.name ?? '').trim()
        const eventType = String(args.event_type ?? 'webhook').trim()
        if (!name) return 'Fehler: name erforderlich.'
        this.db.upsertWebhook(name, args.secret ? String(args.secret) : null, eventType)
        return `Webhook '${name}' erstellt. Aufrufen unter: POST /api/webhooks/${name}`
      }

      case 'delete': {
        const name = String(args.name ?? '').trim()
        if (!name) return 'Fehler: name erforderlich.'
        this.db.deleteWebhook(name)
        return `Webhook '${name}' geloescht.`
      }

      default:
        return `Unbekannte Aktion: ${action}`
    }
  }
}
